### ISSUE TITLE
False positive with function that returns union of `TypeVar`s even when the inferred type seems correct

### ISSUE THREAD

**Issue by Azureblade3808** [microsoft/pyright#6651]
**Describe the bug**
If a function is declared like `def func[T0, T1](arg0: T0, arg1: T1) -> T0 | T1: ...`, and gets called like `func(a, b)` when type of `a` is a union which includes types of `b`, the return type will seem correctly inferred, but cause errors when being assigned with declared type or being used as a generic argument.

**Code or Screenshots**
```python
from collections.abc import Iterable

def padded[T, U](iterable: Iterable[T], fill_value: U, /) -> Iterable[T | U]: ...

a = [int(), str()]
b = int()
c = padded(a, b)
d: Iterable[int | str] = padded(a, b)
e = list(padded(a, b))
```
![image](https://github.com/microsoft/pyright/assets/15389744/16603d90-1588-42e8-81c4-cc14aeed0c36)

**VS Code extension or command-line**
Tested with Pyright version 1.1.338, but the issue also appears in older versions.

**Comment by erictraut** [comment]
This is a very difficult case to handle.

The expression `padding(a, b)` is evaluated using bidirectional type inference. The "expected type" in this case is `Iterable[int | str]`. The return type of `padded` is declared as `Iterable[T | U]`. A set of constraints must be imposed on `T` and `U` prior to evaluating the types of argument expressions `a` and `b`. The problem is that the appropriate constraints are not obvious. There are several possible constraint sets in this case:

1. T <: int, U <: str
2. T <: str, U <: int
3. T <: (int | str), no constraint for U
4. U <: (int | str), no constraint for T
5. T <: (int | str), U <: (int | str)
6. No constraints for either T or U

Choosing any one of these could produce incorrect results. Currently, pyright (and mypy) choose one of the above and proceed under the hope that it works in most cases. In the case of pyright, the choice is based on some heuristics that produce the desired results the vast majority of the time, but they are fallable. Pyright first sorts the subtypes of the unions (so the results don't depend on union ordering). If the number of subtypes in the source union (`int | str`) match the number of unbound type variables in the dest union (`T | U`) it applies the constraints in sorted order, one for each type variable. In this code sample, pyright's heuristic results in the first constraint set in the list above.

I'll note that mypy also generates a similar error for this code. It runs into the same issue.

If PEP 718 is accepted, there would be a straightforward workaround — `padded[int | str, str](a, b)`.

I'll think about this a bit more, but my inclination is to close this as "won't fix". The fixes I can currently think of would likely do more damage than good — i.e. they'd result in more false positives.

**Comment by erictraut** [comment]
I've decided to close the issue as "won't fix".
